Adjudicating eligibility of person during Field sync

jdzak opened the pull request:
This will mark a person as ineligible when during a Field sync and prevents scheduling further events.

I tested this works locally, but haven't written any automated tests because there's not currently an easy way to mock out the interaction with PSC. I can look into doing this, but I have a feeling it's not trivial.

Commit:
Adjudicating eligibility of person during Field sync. #3783

hannahwhy commented on app/models/field/scheduling.rb:
```diff
@@ -14,10 +14,10 @@ def advance_participant_schedules
       # Ensure we have up-to-date data; we're entering this method out of a
       # lot of manipulations.
-      current_participants.each(&:reload)
+      eligible_participants.each(&:reload)
```
The documentation for this module has to be updated.

hannahwhy commented on app/models/merge.rb:
```diff
@@ -169,6 +169,9 @@ def run
             superposition.prepare_for_sync(self)
             superposition.sync_with_psc
 
+            # .. then determine eligibility
```
Ellipses have three periods.  There should be two: one before, one after, with no spaces between the ellipses and words.

hannahwhy commented on app/models/field/protocol_eligibility.rb:
```diff
@@ -0,0 +1,14 @@
+
+      current_participants.each(&:reload)
+
+      ineligible_participants, self.eligible_participants = current_participants.partition{ |p| p.ineligible? }
```
How many queries will `ineligible?` issue?  If it's quite a few, can it be sped up?  At this point you've got access to persisted Participant data, so it might be worth eager-loading all dependent data.

jdzak replied:
There are a couple different code paths, but the path leading to the most queries is PBS (6 queries). It looks like everything that can be eager-loaded is already being eager-loaded.

We could reduce the number of times this is called by only checking eligibility when a screening event exists for a participant. This is also done in the `events_controller.rb` before checking eligibility, so maybe this is worth doing.

hannahwhy commented on app/models/field/protocol_eligibility.rb:
```diff
@@ -0,0 +1,14 @@
+
+      ineligible_participants, self.eligible_participants = current_participants.partition{ |p| p.ineligible? }
+      
+      ineligible_participants.map(&:person).compact.each(&:adjudicate_eligibility)
```
As an example, you can avoid many unnecessary queries here by operating on the association:

``` ruby
ParticipantPersonLink.self_links.where(:participant_id => ineligible_participants).
  includes(:person => whatever_else).map(&:person).each(&:adjudicate_eligibility)
```

There's no `self_links` scope (AFAIK) but one could be written.  This has the benefit of eliminating a `#compact`, too, since you'll never get back a Participant without a Person this way.

Or, push it up a bit higher:

``` ruby
ps = Participant.where(:id => current_participants.map(&:id)).includes(:participant_person_link => :person)  # and whatever else
in, el = ps.partition(&:ineligible?)

in.map(&:person).each(&:adjudicate_eligibility)
```

hannahwhy replied:
Actually, the more I look at that, the more I think this method makes sense:

``` ruby
class Participant
  # Marks a bunch of participants (identified by ID) ineligible.
  def self.disqualify(ids)
  end
end
```

This module could then call that.

jdzak replied:
Sure, I can change this.

Commit:
Doc/comment updates. #3783

Commit:
Simplify code and eager-load person. #3783

hannahwhy commented on app/models/field/protocol_eligibility.rb:
```diff
@@ -0,0 +1,14 @@
+module Field
+  module ProtocolEligibility
+
```
This blank line shouldn't be here.

hannahwhy commented on app/models/field/protocol_eligibility.rb:
```diff
@@ -0,0 +1,14 @@
+module Field
+  module ProtocolEligibility
```
What is this module's purpose?

jdzak replied:
Wrapper around Participant::adjudicate_participants_and_disqualify_ineligible so Merge#run can find/delete ineligible participants.

hannahwhy replied:
That, or something like it, should be this module's documentation.

Commit:
Changed EligibilityAdjudicator to be included on Participant and more. #3783

More..
1. Changed ::adjudicate_eligibility to only group participants by eligible or ineligible
2. Added ::disqualify method to remove participant(s) from the study
3. Added ::adjudicate_eligibility_and_disqualify_ineligible to do both of above

Commit:
Removed blank line. #3783

Commit:
Outdated comment..yoink. #3783

Commit:
Trailing elipses. #3783

Commit:
Fix class name referenced. #3783

Commit:
Document purpose of ProtocolEligibility module. #3783

jdzak commented:
Alright, I have addressed all comments, ready for review again.

Major changes to note since the initial pull request notification.
1. Changed EligibilityAdjudicator to be a module that is included on Participant
2. Changed ::adjudicate_eligibility to ONLY group participants by eligible or ineligible (not make ineligible)
3. Renamed ::make_ineligible to ::disqualify
4. Added eager-loading of associations to ::disqualify
5. Added ::adjudicate_eligibility_and_disqualify_ineligible as a convenience method for adjudicating eligibility and disqualifying ineligible participants

Commit:
Reload event so redirec logic properly works. #3783

pfriedman commented:
Merged PR into develop
